lib/handler/decorator-handler.js

```diff
@@ -14,12 +26,88 @@ module.exports = class DecoratorHandler {
     this.#handler = handler
   }
 
+  // New API
+
+  onRequestStart (abort) {
+    this.#handler.onRequestStart?.(abort)
+    this.#handler.onConnect?.(abort)
+  }
+
+  onResponseStart (resume) {
+    let ret = true
+
+    if (this.#handler.onResponseStart?.(resume) === false) {
+      ret = false
+    }
+
+    this.#resume = resume
+
+    return ret
+  }
+
+  onResponseHeaders (headers, statusCode) {
+    let ret = true
+
+    if (this.#handler.onResponseHeaders?.(headers, statusCode) === false) {
+      ret = false
+    }
+
+    if (this.#handler.onHeaders) {
```

Comment:
Shall we prevent to have both the old and new API?

Reply:
How would we do that?

Reply:
Maybe we can check at [`dispatcher-base`](https://github.com/nodejs/undici/blob/e6e87c13dbf8788fe158fd60631cee78eb84f198/lib/dispatcher/dispatcher-base.js#L128) when calling `dispatch`. We already validate the `handler`, we can take advantage of validating the method as well if we want to avoid receiving both functions/method at the same time.